crypto/init.c

```diff
@@ -474,12 +474,12 @@ void OPENSSL_cleanup(void)
                     "RAND_cleanup()\n");
 
 #endif
-    CRYPTO_cleanup_all_ex_data();
-    EVP_cleanup();
-    CONF_modules_free();
 #ifndef OPENSSL_NO_ENGINE
     ENGINE_cleanup();
 #endif
+    CRYPTO_cleanup_all_ex_data();
+    EVP_cleanup();
+    CONF_modules_free();
```

Comment:
Note that I just ordered this after the debug lines above, which I assumed carried the original intent. The only constraint that I needed to satisfy was "engine cleanup before crypto ex data cleanup" - please send a PR that satisfies any addl requirements (and reorder the debug lines!).